File: libs/core/src/symbols/named_type_symbol.cpp
/**
 * @file named_type_symbol.cpp
 * @author Francesco Corso
 * @date 8/8/2026
 * @brief
 */
module prism.core:symbols.named_type_symbol.impl;

import :symbols.named_type_symbol;

namespace prism
{
    void NamedTypeSymbol::write_display_string(TextWriter &writer) const
    {
        switch (special_type())
        {
            case SpecialType::void_:
                writer.write("void");
                break;
            case SpecialType::bool_:
                writer.write("bool");
                break;
            case SpecialType::i8:
                writer.write("i8");
                break;
            case SpecialType::i16:
                writer.write("i16");
                break;
            case SpecialType::i32:
                writer.write("i32");
                break;
            case SpecialType::i64:
                writer.write("i64");
                break;
            case SpecialType::i128:
                writer.write("i128");
                break;
            case SpecialType::isize:
                writer.write("isize");
                break;
            case SpecialType::u8:
                writer.write("u8");
                break;
            case SpecialType::u16:
                writer.write("u16");
                break;
            case SpecialType::u32:
                writer.write("u32");
                break;
            case SpecialType::u64:
                writer.write("u64");
                break;
            case SpecialType::u128:
                writer.write("u128");
                break;
            case SpecialType::usize:
                writer.write("usize");
                break;
            case SpecialType::f16:
                writer.write("f16");
                break;
            case SpecialType::f32:
                writer.write("f32");
                break;
            case SpecialType::f64:
                writer.write("f64");
                break;
            case SpecialType::char_:
                writer.write("char");
                break;
            case SpecialType::char16:
                writer.write("char16");
                break;
            case SpecialType::rune:
                writer.write("rune");
                break;
            case SpecialType::str:
                writer.write("str");
                break;
            default:
                writer.write(name().as_string_view());
        }
    }
} // namespace prism

File: src/Prism.Core/Symbols/NamedTypeSymbol.cs
namespace Prism.Core.Symbols;

public abstract class NamedTypeSymbol : TypeSymbol
{
    private protected NamedTypeSymbol(
        string name,
        Symbol? containingSymbol,
        SpecialType specialType = SpecialType.None
    )
        : base(name, containingSymbol, specialType) { }

    public sealed override void WriteDisplayString(TextWriter writer)
    {
        switch (SpecialType)
        {
            case SpecialType.Void:
                writer.Write("void");
                break;
            case SpecialType.Bool:
                writer.Write("bool");
                break;
            case SpecialType.I8:
                writer.Write("i8");
                break;
            case SpecialType.I16:
                writer.Write("i16");
                break;
            case SpecialType.I32:
                writer.Write("i32");
                break;
            case SpecialType.I64:
                writer.Write("i64");
                break;
            case SpecialType.I128:
                writer.Write("i128");
                break;
            case SpecialType.ISize:
                writer.Write("isize");
                break;
            case SpecialType.U8:
                writer.Write("u8");
                break;
            case SpecialType.U16:
                writer.Write("u16");
                break;
            case SpecialType.U32:
                writer.Write("u32");
                break;
            case SpecialType.U64:
                writer.Write("u64");
                break;
            case SpecialType.U128:
                writer.Write("u128");
                break;
            case SpecialType.USize:
                writer.Write("usize");
                break;
            case SpecialType.F32:
                writer.Write("f32");
                break;
            case SpecialType.F64:
                writer.Write("f64");
                break;
            case SpecialType.Char:
                writer.Write("char");
                break;
            case SpecialType.Char16:
                writer.Write("char16");
                break;
            case SpecialType.Rune:
                writer.Write("rune");
                break;
            case SpecialType.Str:
                writer.Write("str");
                break;
            case SpecialType.None:
            default:
            {
                if (ContainingNamespace is { IsGlobal: false } container)
                {
                    container.WriteDisplayString(writer);
                    writer.Write("::");
                }

                writer.Write(Name);
                break;
            }
        }
    }
}
